jfr/src/main/java/io/netty/handler/jfr/JfrEventHandler.java

```diff
@@ -0,0 +1,358 @@
+//            event.setChannel(ctx.channel());
+//            event.commit();
+//        }
+//    }
+
+    @Override
+    public void write(ChannelHandlerContext ctx, Object msg, ChannelPromise promise) throws Exception {
+        WriteEvent event = new WriteEvent();
+        event.begin();
+        ctx.write(msg, promise);
+        if (event.shouldCommit()) {
+            event.setChannel(ctx.channel());
+            event.msg = String.valueOf(msg);
+            event.commit();
+        }
+    }
+
+    @Override
+    public void channelWritabilityChanged(ChannelHandlerContext ctx) throws Exception {
+        ChannelWritabilityChangedEvent event = new ChannelWritabilityChangedEvent();
+        event.begin();
+        ctx.fireChannelWritabilityChanged();
+        if (event.shouldCommit()) {
+            event.setChannel(ctx.channel());
+            event.isWritable = ctx.channel().isWritable();
+            event.commit();
+        }
+    }
+
+    @Override
+    public void flush(ChannelHandlerContext ctx) throws Exception {
+        FlushEvent event = new FlushEvent();
+        event.begin();
+        ctx.flush();
+        if (event.shouldCommit()) {
+            event.setChannel(ctx.channel());
+            event.commit();
+        }
+    }
+
+    // Event types
+
+    @Category({ "Netty", "ChannelHandler" })
+    private abstract static class ChannelEvent extends Event {
+        @Label("Channel")
+        String channel;
+        @Label("Channel ID")
+        String channelId;
+
+        void setChannel(Channel channel) {
+            this.channel = String.valueOf(channel);
+            //TODO: More specific channel fields?
+            this.channelId = channel.id().asLongText();
+        }
+    }
+
+    @Label("Bind")
+    @Name("io.netty.channelhandler.bind")
+    private static final class BindEvent extends ChannelEvent {
+        @Label("Local address")
+        String localAddress;
+    }
+
+    @Label("Write")
+    @Name("io.netty.channelhandler.write")
+    private static final class WriteEvent extends ChannelEvent {
+        @Label("Message")
+        String msg;
+    }
+
+    @Label("Channel Read")
+    @Name("io.netty.channelhandler.channelread")
+    private static final class ChannelReadEvent extends ChannelEvent {
+        @Label("Message")
+        String msg;
+    }
+
+    @Label("Channel Read Complete")
+    @Name("io.netty.channelhandler.channelreadcomplete")
+    @Enabled(false)
+    private static final class ChannelReadCompleteEvent extends ChannelEvent {
+    }
+
+    @Label("Close")
+    @Name("io.netty.channelhandler.close")
+    private static final class CloseEvent extends ChannelEvent {
+    }
+
+    @Label("Disconnect")
+    @Name("io.netty.channelhandler.disconnect")
+    private static final class DisconnectEvent extends ChannelEvent {
+    }
+
+    @Label("Connect")
+    @Name("io.netty.channelhandler.connect")
+    private static final class ConnectEvent extends ChannelEvent {
+        @Label("Remote address")
```

Comment:
@Label should use headline-style capitalization (same as Eclipse), i.e "Remote Address"